src/audio/google_audio_post_processing.c

```diff
@@ -0,0 +1,551 @@
+	case SOF_CTRL_CMD_BINARY:
+		return gapp_ctrl_set_bin_data(dev, cdata);
+	default:
+		comp_err(dev, "gapp_ctrl_set_data(): invalid cdata->cmd");
+		return -EINVAL;
+	}
+}
+
+static int gapp_get_comp_config(struct comp_data *cd,
+				struct sof_ipc_ctrl_data *cdata,
+				int max_data_size)
+{
+	if (sizeof(cd->config) > max_data_size)
+		return -EINVAL;
+
+	memcpy_s(cdata->data->data, max_data_size, cd->config,
+		 sizeof(cd->config));
+	cdata->data->abi = SOF_ABI_VERSION;
+	cdata->data->size = sizeof(cd->config);
+
+	return 0;
+}
+
+static int gapp_get_internal_config(struct comp_data *cd,
+				    struct sof_ipc_ctrl_data *cdata,
+				    int max_data_size)
+{
+	int data_size;
+	int blob_size = cdata->num_elems;
+
+	if (blob_size > max_data_size)
+		return -EINVAL;
+
+	data_size = GoogleAudioPostProcessingGetConfig(
+		cd->state, cdata->data->type, cdata->msg_index,
+		(uint8_t *)cdata->data->data, blob_size);
+
+	if (data_size < 0)
+		return -EINVAL;
+
+	cdata->data->abi = SOF_ABI_VERSION;
+	cdata->data->size = data_size;
+
+	return 0;
+}
+
+static int gapp_ctrl_get_bin_data(struct comp_dev *dev,
+				  struct sof_ipc_ctrl_data *cdata,
+				  int max_data_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	assert(cd);
+	comp_dbg(
+		dev,
+		"gapp_ctrl_get_bin_data(): type = %u, comp_id = %u, index = %u, blob_size = %d",
+		cdata->data->type, cdata->comp_id, cdata->msg_index,
+		cdata->num_elems);
+
+	switch (cdata->data->type) {
+	case SOF_GAPP_CONFIG:
+		return gapp_get_comp_config(cd, cdata, max_data_size);
+	case SOF_GAPP_MODEL:
+		COMPILER_FALLTHROUGH
+	case SOF_GAPP_DATA:
+		return gapp_get_internal_config(cd, cdata, max_data_size);
+	default:
+		comp_err(dev,
+			 "gapp_ctrl_get_bin_data(): unknown binary data type");
+		return -EINVAL;
+	}
+}
+
+static int gapp_ctrl_get_data(struct comp_dev *dev,
+			      struct sof_ipc_ctrl_data *cdata,
+			      int max_data_size)
+{
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		return gapp_ctrl_get_bin_data(dev, cdata, max_data_size);
+	default:
+		comp_err(dev, "smart_amp_ctrl_get_data(): invalid cdata->cmd");
+		return -EINVAL;
+	}
+}
+
+static int gapp_cmd(struct comp_dev *dev, int cmd, void *data,
+		    int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	switch (cmd) {
+	case COMP_CMD_SET_VALUE:
+		return gapp_ctrl_set_val(dev, cdata);
+	case COMP_CMD_GET_VALUE:
+		return gapp_ctrl_get_val(dev, cdata, max_data_size);
+	case COMP_CMD_SET_DATA:
+		return gapp_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return gapp_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		comp_err(dev, "gapp_cmd(): unhandled command %d", cmd);
+		return -EINVAL;
+	}
+}
+
+static struct comp_dev *gapp_new(const struct comp_driver *drv,
+				 struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *gapp;
+	struct sof_ipc_comp_process *ipc_gapp =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_data *cd;
+	int ret = 0;
```

Comment:
looks like init is not required

Reply:
I am not sure about this comment, do you mean the whole prepare() function is not required (cuz there is no init in the highlighted lines), but in prepare() we check values and do setup/configuration, do you mean all should be moved to new()?

Reply:
no I meant ret doesnt need to be initialized to 0

Reply:
I see. fixed.